### DISCUSSION TITLE
Doesn't work with custom parser

### DISCUSSION THREAD

**Discussion by so1ve** [eslint/markdown#221] [Q&A]
I don't think a reproduction is needed for this issue :(

For example, in eslint config:

```js
{
  extends: ["plugin:markdown/recommended"],
  overrides: [
    { files: ["*.md"], parser: "eslint-parser-plain", rules: {} }
  ]
}
```

The parser does not take effect. (Actually it doesn't run at all!)

Is this by design?

**Comment by sam3k** [comment]
Hi @so1ve, could you provide a reproducible demo? Taking a quick look at the plugin [docs](https://github.com/eslint/eslint-plugin-markdown) shows that the API has changed between v1 and v2. Version 2 is configured like this it seems:

```javascript
module.exports = {
    extends: "plugin:markdown/recommended",
    plugins: ["markdown"],
    overrides: [
        {
            files: ["**/*.md"],
            processor: "markdown/markdown"
        },
    ]
};
```

**Comment by so1ve** [comment]
Sure, please checkout https://github.com/so1ve/eslint-plugin-markdown-repro

1. Install the dependencies and run `npm run lint`
2. Console logs the content of test.md
3. Comment out the line in `.eslintrc.js`
4. Run `npm run lint` again
5. Console logs nothing

**Comment by sam3k** [comment]
Thank you for the demo @so1ve. Just ran it on my machine and I see that the linter throws error if you use `markdown/markdown` as processor:

```javascript
module.exports = {
  parserOptions: {
    ecmaVersion: "latest",
    sourceType: "module"
  },
  plugins: ["markdown"],
  overrides: [
    {
      files: ["**/*.md"],
      processor: "markdown/markdown",
    },
  ],
  rules: {
    semi: "error"
  }
};
```

It also works if you remove the processor:

```javascript
module.exports = {
  parserOptions: {
    ecmaVersion: "latest",
    sourceType: "module"
  },
  plugins: ["markdown"],
  overrides: [
    {
      files: ["**/*.md"],
    },
  ],
  rules: {
    semi: "error"
  }
};
```

Now, if I understood you correctly, it seems like the markdown errors stop working once you add a custom `parser.js`. I am able to get it to still throw error if I have the following config:

```javascript
module.exports = {
  parserOptions: {
    ecmaVersion: "latest",
    sourceType: "module"
  },
  plugins: ["markdown"],
  overrides: [
    {
      files: ["**/*.md"],
      processor: "markdown/markdown", // Comment out this line
      parser: require.resolve("./parser.js")
    },
  ],
  rules: {
    semi: "error"
  }
};
```

and then use the following format in `parser.js` which is based on the example given in the [docs](https://eslint.org/docs/latest/extend/custom-parsers#example):

```javascript
var espree = require("espree");

function parseForESLint(code, options) {
  console.log('parseForESLint');
  return {
    ast: espree.parse(code, options),
    services: {
      isPlain: true,
      foo: function() {
        console.log("foo");
      }
    },
    scopeManager: null,
    visitorKeys: null
  };
};

module.exports = { parseForESLint };
```

This calls the custom parser and throws the error still.

**Comment by nzakas** [comment]
@sam3k when you say "throws errors", are you saying that ESLint is reporting linting errors or that it is throwing an error (with a stack trace)?

**Comment by sam3k** [comment]
@nzakas I meant ESLint reports lint error:


![Image](https://github.com/eslint/eslint-plugin-markdown/assets/180152/98010af3-6d9a-4b80-a15a-e28f716c1fa4)

**Comment by sam3k** [comment]
@so1ve were you able to try out the configuration I've suggested?

**Comment by so1ve** [comment]
@sam3k Ah, sorry. Before I was busy with exams and didn't have time to see github. I'll try your solution later

**Comment by so1ve** [comment]
Haven't tried it yet, but I think it won't work as my expectations. In your code, you call `espree.parse` to parse the code, which means the input is JavaScript code, not markdown. However, I need a custom parser to return a `Program` node that contains the markdown file's content to let prettier format it

**Comment by sam3k** [comment]
@so1ve it seems like it returns JS AST. I've seen the `ast` object being defined as a Program like the following but not sure if this can be of any help to you:

```javascript
function parseForESLint(code, options) {
  return {
    ast: {
      // ast is JS ast. We don't have JS, so this AST is for empty JS file
      type: 'Program',
      start: 0,
      end: 0,
      loc: { start: { line: 1, column: 0 }, end: { line: 1, column: 0 } },
      range: [0, 0],
      body: [],
      tokens: [],
      comments: [],
      // Used only by eslint-plugin-markdown-language
      mdCode: code,
    },
    services: {
      isPlain: true,
      foo: function() {
        console.log("foo");
      }
    },
    scopeManager: null,
    visitorKeys: null
  };
};

module.exports = { parseForESLint };
```

Not sure if this is helpful; but could you give it a shot?

**Comment by so1ve** [comment]
Maybe I am doing something wrong?

```js
function parseForESLint(code, options) {
  console.log(code) // This logs nothing :(
  return {
    ast: {
      // ast is JS ast. We don't have JS, so this AST is for empty JS file
      type: 'Program',
      start: 0,
      end: 0,
      loc: { start: { line: 1, column: 0 }, end: { line: 1, column: 0 } },
      range: [0, 0],
      body: [],
      tokens: [],
      comments: [],
      // Used only by eslint-plugin-markdown-language
      mdCode: code,
    },
    services: {
      isPlain: true,
      foo: function() {
        console.log("foo");
      }
    },
    scopeManager: null,
    visitorKeys: null
  };
};

module.exports = { parseForESLint };
```

**Comment by sam3k** [comment]
@so1ve everything looks good.  Can you try [this fork ](https://github.com/so1ve/eslint-plugin-markdown-repro/commit/033124a86f5cd003a672f2d465d6a68035c964c5) I've made?

**Comment by so1ve** [comment]
It doesn't work either... It logs nothing :(

**Comment by nzakas** [comment]
It looks like this isn't about a problem with the plugin, but rather, a question about how to configure ESLint, so I'm moving this to a discussion.
